cpp/src/lists/contains.cu

```diff
@@ -190,129 +151,91 @@ struct search_list_nested_types_fn {
-                 "Type/Scale of search key does not match list column element type.",
-                 cudf::data_type_error);
-    CUDF_EXPECTS(search_keys.type().id() != type_id::EMPTY, "Type cannot be empty.");
+  CUDF_EXPECTS(child.type() == search_keys.type(),
+               "Type/Scale of search key does not match list column element type.",
+               cudf::data_type_error);
+  CUDF_EXPECTS(search_keys.type().id() != type_id::EMPTY, "Type cannot be empty.");
 
-    auto const search_keys_have_nulls = search_keys.has_nulls();
+  auto const search_keys_have_nulls = search_keys.has_nulls();
 
-    auto const num_rows = lists.size();
+  auto const num_rows = lists.size();
 
-    auto const lists_cdv_ptr = column_device_view::create(lists.parent(), stream);
-    auto const input_it      = cudf::detail::make_counting_transform_iterator(
-      size_type{0},
-      [lists = cudf::detail::lists_column_device_view{*lists_cdv_ptr}] __device__(auto const idx) {
-        return list_device_view{lists, idx};
-      });
+  auto const lists_cdv_ptr = column_device_view::create(lists.parent(), stream);
+  auto const input_it      = cudf::detail::make_counting_transform_iterator(
+    size_type{0},
+    [lists = cudf::detail::lists_column_device_view{*lists_cdv_ptr}] __device__(auto const idx) {
+      return list_device_view{lists, idx};
+    });
+
+  auto out_positions = make_numeric_column(
+    data_type{type_to_id<size_type>()}, num_rows, cudf::mask_state::UNALLOCATED, stream, mr);
+  auto const output_it = out_positions->mutable_view().template begin<size_type>();
 
-    auto out_positions = make_numeric_column(
-      data_type{type_to_id<size_type>()}, num_rows, cudf::mask_state::UNALLOCATED, stream, mr);
-    auto const output_it = out_positions->mutable_view().template begin<size_type>();
-
-    if constexpr (not cudf::is_nested<Element>()) {
-      index_of_non_nested_types<Element>(
-        input_it, num_rows, output_it, search_keys, search_keys_have_nulls, find_option, stream);
-    } else {  // list + struct
-      index_of_nested_types(input_it, num_rows, output_it, child, search_keys, find_option, stream);
-    }
-
-    if (search_keys_have_nulls || lists.has_nulls()) {
-      auto [null_mask, null_count] = cudf::detail::valid_if(
-        output_it,
-        output_it + num_rows,
-        [] __device__(auto const idx) { return idx != NULL_SENTINEL; },
-        stream,
-        mr);
-      out_positions->set_null_mask(std::move(null_mask), null_count);
-    }
-    return out_positions;
+  auto const keys_tview  = cudf::table_view{{search_keys}};
+  auto const child_tview = cudf::table_view{{child}};
+  auto const has_nulls   = has_nested_nulls(child_tview) || has_nested_nulls(keys_tview);
+  auto const comparator =
+    cudf::experimental::row::equality::two_table_comparator(child_tview, keys_tview, stream);
+  if (search_keys.type().id() == cudf::type_id::STRUCT or
+      search_keys.type().id() == cudf::type_id::LIST) {
```

Comment:
Use `is_nested` rather than explicitly listing the nested types. https://github.com/rapidsai/cudf/blob/dcc8d9195d62c64f0c132d0b67f9abf17207305e/cpp/src/utilities/traits.cpp#L352
```suggestion
  if (cudf::is_nested(search_keys.type())) {
```